File: cli/internal/sse/sse_test.go
package sse

import "testing"

func TestFeed_LeavesPartialEvent(t *testing.T) {
	events, leftover := Feed("event: stage\ndata: \"hi\"\n\nevent: ou")
	if len(events) != 1 {
		t.Fatalf("events: %d", len(events))
	}
	if events[0].Event != "stage" {
		t.Fatalf("event: %q", events[0].Event)
	}
	if leftover != "event: ou" {
		t.Fatalf("leftover: %q", leftover)
	}
}

func TestFeed_DecodesJSONData(t *testing.T) {
	events, _ := Feed(`event: stage
data: {"stage":"turn","state":"done"}

`)
	if len(events) != 1 {
		t.Fatalf("events: %d", len(events))
	}
	m, ok := events[0].Data.(map[string]any)
	if !ok {
		t.Fatalf("data not map: %T", events[0].Data)
	}
	if m["stage"] != "turn" || m["state"] != "done" {
		t.Fatalf("data: %v", m)
	}
}

func TestFeed_FallsBackToRawString(t *testing.T) {
	events, _ := Feed("data: not json\n\n")
	if len(events) != 1 {
		t.Fatalf("events: %d", len(events))
	}
	s, ok := events[0].Data.(string)
	if !ok || s != "not json" {
		t.Fatalf("data: %v", events[0].Data)
	}
}

func TestFeed_IgnoresHeartbeat(t *testing.T) {
	events, _ := Feed(": keep-alive\n\n")
	if len(events) != 0 {
		t.Fatalf("events: %d", len(events))
	}
}

func TestFeed_ParsesIDAsInt(t *testing.T) {
	events, _ := Feed("id: 42\nevent: x\ndata: ok\n\n")
	if len(events) != 1 {
		t.Fatalf("events: %d", len(events))
	}
	if events[0].ID != 42 {
		t.Fatalf("id: %d", events[0].ID)
	}
}

File: sdk/typescript/test/sse.test.ts
import { test, describe } from "node:test";
import assert from "node:assert/strict";
import { parseSse } from "../src/sse.ts";
import { TurnFollower } from "../src/turn.ts";
import type { LogEvent } from "../src/types.ts";

/** Feed a body to the parser in exactly the chunks given, TCP-style. */
function bodyOf(chunks: string[]): ReadableStream<Uint8Array> {
  const encoder = new TextEncoder();
  return new ReadableStream({
    start(controller) {
      for (const chunk of chunks) controller.enqueue(encoder.encode(chunk));
      controller.close();
    },
  });
}

async function collect(chunks: string[]) {
  const out = [];
  for await (const message of parseSse(bodyOf(chunks))) out.push(message);
  return out;
}

describe("parseSse", () => {
  test("reads id, event and data", async () => {
    const messages = await collect(['id: 7\nevent: output\ndata: {"kind":"output"}\n\n']);
    assert.deepEqual(messages, [{ id: "7", event: "output", data: '{"kind":"output"}' }]);
  });

  test("a message split across packets is still one message", async () => {
    // The failure this guards against is not hypothetical: Fountain's replay
    // buffer cuts at a byte count, so a message arriving in two pieces is the
    // normal case, not the edge one.
    const messages = await collect(["id: 1\neve", "nt: output\ndata: {\"a\":", '1}\n\n']);
    assert.equal(messages.length, 1);
    assert.deepEqual(messages[0], { id: "1", event: "output", data: '{"a":1}' });
  });

  test("heartbeat comments are not messages", async () => {
    const messages = await collect([": heartbeat\n\n", "id: 2\nevent: stage\ndata: {}\n\n"]);
    assert.equal(messages.length, 1);
    assert.equal(messages[0]?.id, "2");
  });

  test("multi-line data joins with newlines", async () => {
    const messages = await collect(["data: one\ndata: two\n\n"]);
    assert.equal(messages[0]?.data, "one\ntwo");
  });

  test("CRLF framing works too", async () => {
    const messages = await collect(["id: 3\r\nevent: output\r\ndata: hi\r\n\r\n"]);
    assert.deepEqual(messages[0], { id: "3", event: "output", data: "hi" });
  });

  test("a final message with no trailing blank line is not dropped", async () => {
    const messages = await collect(["id: 9\nevent: stage\ndata: {}"]);
    assert.equal(messages[0]?.id, "9");
  });
});

const stage = (state: string, meta: Record<string, unknown>): LogEvent => ({
  id: 1,
  kind: "stage",
  stage: "turn",
  state,
  data: JSON.stringify(meta),
});

const output = (blocks: unknown[], turnId: string, stream = "acp"): LogEvent => ({
  id: 2,
  kind: "output",
  stream,
  turn_id: turnId,
  blocks: blocks as LogEvent["blocks"],
});

describe("TurnFollower", () => {
  test("ignores a different turn's lifecycle", () => {
    const follower = new TurnFollower(2);
    follower.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    assert.equal(follower.started, false);
    follower.apply(stage("done", { turn_number: 1, turn_id: "t1" }));
    assert.equal(follower.finished, false);
  });

  test("thinking is reported but is not the answer", () => {
    const follower = new TurnFollower(1);
    follower.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    const events = follower.apply(output([{ kind: "thinking", body: "hmm" }], "t1"));
    assert.ok(events.some((e) => e.type === "thinking"));
    assert.equal(follower.text, "");
  });

  test("a result block is the answer only when nothing else was said", () => {
    const follower = new TurnFollower(1);
    follower.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    follower.apply(output([{ kind: "result", body: "exit 0" }], "t1"));
    assert.equal(follower.text, "exit 0");

    const chatty = new TurnFollower(1);
    chatty.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    chatty.apply(output([{ kind: "text", body: "All set." }], "t1"));
    chatty.apply(output([{ kind: "result", body: "exit 0" }], "t1"));
    assert.equal(chatty.text, "All set.");
  });

  test("a tool is recorded once, in first-use order", () => {
    const follower = new TurnFollower(1);
    follower.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    follower.apply(output([{ kind: "tool_use", name: "Bash" }], "t1"));
    follower.apply(output([{ kind: "tool_use", name: "Read" }], "t1"));
    follower.apply(output([{ kind: "tool_use", name: "Bash" }], "t1"));
    assert.deepEqual(follower.toolsUsed, ["Bash", "Read"]);
  });

  test("the terminal stage carries how it ended", () => {
    const follower = new TurnFollower(1);
    follower.apply(stage("started", { turn_number: 1, turn_id: "t1" }));
    follower.apply(stage("failed", { turn_number: 1, turn_id: "t1", exit_code: 2, reason: "oom" }));
    assert.equal(follower.state, "failed");
    assert.equal(follower.exitCode, 2);
    assert.equal(follower.reason, "oom");
  });

  test("matches on turn_id once known, even if the turn number drifts", () => {
    const follower = new TurnFollower(3);
    follower.apply(stage("started", { turn_number: 3, turn_id: "t3" }));
    follower.apply(output([{ kind: "text", body: "hi" }], "t3"));
    // A resumed sandbox renumbering its turns must not orphan the follower.
    follower.apply(stage("done", { turn_number: 99, turn_id: "t3" }));
    assert.equal(follower.state, "done");
    assert.equal(follower.text, "hi");
  });
});
